Objective-C/CBLDatabase.mm

```diff
@@ -0,0 +1,373 @@
+//
+//  CBLDatabase.m
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 12/15/16.
+//  Copyright © 2016 Couchbase. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//    http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+
+#import "CBLDatabase.h"
+#import "CBLInternal.h"
+#import "CBLDocument.h"
+#import "CBLCoreBridge.h"
+#import "CBLStringBytes.h"
+#import "CBLMisc.h"
+
+
+@implementation CBLDatabaseOptions
+
+
+- (instancetype) copyWithZone:(NSZone *)zone {
+    CBLDatabaseOptions* o = [[self.class alloc] init];
+    o.directory = self.directory;
+    o.encryptionKey = self.encryptionKey;
+    o.readOnly = self.readOnly;
+    return o;
+}
+
+
++ (instancetype) defaultOptions {
+    return [[CBLDatabaseOptions alloc] init];
+}
+
+
+@end
+
+
+@implementation CBLDatabase {
+    NSString* _name;
+    CBLDatabaseOptions* _options;
+    C4Database* _c4db;
+    NSMapTable<NSString*, CBLDocument*>* _documents;
+    NSMutableSet<CBLDocument*>* _unsavedDocuments;
+}
+
+
+@synthesize c4db=_c4db;
+
+
+static const C4DatabaseConfig kDBConfig = {
+    .flags = (kC4DB_Create | kC4DB_AutoCompact | kC4DB_Bundled | kC4DB_SharedKeys),
+    .storageEngine = kC4SQLiteStorageEngine,
+    .versioning = kC4RevisionTrees,
+};
+
+
+static void logCallback(C4LogDomain domain, C4LogLevel level, C4Slice message) {
+    static const char* klevelNames[5] = {"Debug", "Verbose", "Info", "WARNING", "ERROR"};
+    NSLog(@"CouchbaseLite %s %s: %.*s", c4log_getDomainName(domain), klevelNames[level],
+          (int)message.size, (char*)message.buf);
+}
+
+
++ (void) initialize {
+    if (self == [CBLDatabase class]) {
+        c4log_register(kC4LogWarning, &logCallback);
+    }
+}
+
+
+- (instancetype) initWithName: (NSString*)name
+                        error: (NSError**)outError {
+    return [self initWithName: name
+                      options: [CBLDatabaseOptions defaultOptions]
+                        error: outError];
+}
+
+
+- (instancetype) initWithName: (NSString*)name
+                      options: (CBLDatabaseOptions*)options
+                        error: (NSError**)outError {
+    _name = name;
```

Comment:
Just noticed this — need to call `[super init]` and check that for nil, as usual in an `init` method.